ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1764,95 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+BOOLEAN
+IopValidateID(
+    _In_ PWCHAR Id,
+    _In_ BUS_QUERY_ID_TYPE QueryType)
+{
+    PWCHAR PtrChar;
+    PWCHAR StringEnd;
+    WCHAR Char;
+    ULONG SeparatorsCount = 0;
+    PWCHAR PtrPrevChar = NULL;
+    ULONG MaxSeparators;
+    BOOLEAN IsMultiSz;
+
+    PAGED_CODE();
+
+    switch (QueryType)
+    {
+        case BusQueryDeviceID:
+            MaxSeparators = MAX_SEPARATORS_DEVICEID;
+            IsMultiSz = FALSE;
+            break;
+        case BusQueryInstanceID:
+            MaxSeparators = MAX_SEPARATORS_INSTANCEID;
+            IsMultiSz = FALSE;
+            break;
+
+        case BusQueryHardwareIDs:
+        case BusQueryCompatibleIDs:
+            IsMultiSz = TRUE;
+            break;
+
+        default:
+            DPRINT1("IopValidateID: Not handled QueryType - %x\n", QueryType);
+            return FALSE;
+    }
+
+    StringEnd = Id + MAX_DEVICE_ID_LEN;
+
+    for (PtrChar = Id; PtrChar < StringEnd; PtrChar++)
+    {
+        Char = *PtrChar;
+
+        if (Char == UNICODE_NULL)
+        {
+            if (!IsMultiSz || (PtrPrevChar && PtrChar == PtrPrevChar + 1))
+            {
+                if (MaxSeparators == SeparatorsCount || IsMultiSz)
+                {
+                    return TRUE;
+                }
+
+                DPRINT1("IopValidateID: SeparatorsCount - %lu, MaxSeparators - %lu\n",
+                        SeparatorsCount, MaxSeparators);
+                goto ErrorExit;
+            }
+
+            StringEnd = PtrChar + MAX_DEVICE_ID_LEN + 1;
+            PtrPrevChar = PtrChar;
+        }
+        else if (Char < ' ' || Char > 0x7F || Char == ',')
+        {
+            DPRINT1("IopValidateID: Invalid character - %04X\n", Char);
+            goto ErrorExit;
+        }
+        else if (Char == ' ')
```

Comment:
Is it possible to test on Windows what happens when a pnp driver wants to report IDs with spaces in them?

Reply:
No problem :)
![usbhub_new-replace-revision-with-spaces-usbh_pdoqueryid-busqueryhardwareids](https://user-images.githubusercontent.com/17290914/35680728-812d5be8-079e-11e8-97d6-09cd8660af90.png)
![virtualbox_winxp_defrag_noswap_01_02_2018_21_19_40](https://user-images.githubusercontent.com/17290914/35680759-8dbf6e82-079e-11e8-9d99-c1f0695bd267.png)

Reply:
I see :) It would be interesting also to know what happens on e.g. windows 7 for this precise case (at least just to know what's happening: whether the doc would then be correct etc...)

Reply:
I do not have 7 in VBox :)